tools/clang/include/clang/Basic/DiagnosticSemaKinds.td

```diff
@@ -7802,6 +7802,12 @@ def err_hlsl_logical_binop_scalar : Error<
    "operands for short-circuiting logical binary operator must be scalar, for non-scalar types use '%select{and|or}0'">;
 def err_hlsl_ternary_scalar : Error<
    "condition for short-circuiting ternary operator must be scalar, for non-scalar types use 'select'">;
+def err_hlsl_no_recursion : Error<
+   "recursive functions are not allowed: entry (or export) function calls recursive function '%0'">;
+def err_hlsl_no_recursion_via_patch : Error<
+   "recursive functions are not allowed: patch constant function calls recursive function '%0'">;
```

Comment:
Can we merge these two diagnostics and use a select instead?

```suggestion
def err_hlsl_no_recursion : Error<
   "recursive functions are not allowed: %select{entry|export|patch constant}1 function calls recursive function '%0'">;
```

Then you can just add `0`, `1`, or `2` to select between “entry”, “export” or “patch constant” as appropriate.